yarn-project/prover-client/src/orchestrator/orchestrator.ts

```diff
@@ -0,0 +1,515 @@
+    private prover: RollupProver,
+  ) {
+    this.simulator = new RealRollupCircuitSimulator(simulationProvider);
+  }
+
+  public static new(db: MerkleTreeOperations, simulationProvider: SimulationProvider, prover: RollupProver) {
+    const orchestrator = new ProvingOrchestrator(db, simulationProvider, getVerificationKeys(), prover);
+    orchestrator.start();
+    return Promise.resolve(orchestrator);
+  }
+
+  public start() {
+    this.jobProcessPromise = this.processJobQueue();
+  }
+
+  public async stop() {
+    this.stopped = true;
+    this.jobQueue.cancel();
+    await this.jobProcessPromise;
+  }
+
+  /**
+   * Starts off a new block
+   * @param numTxs - The number of real transactions in the block
+   * @param globalVariables - The global variables for the block
+   * @param l1ToL2Messages - The l1 to l2 messages for the block
+   * @param emptyTx - The instance of an empty transaction to be used to pad this block
+   * @returns A proving ticket, containing a promise notifying of proving completion
+   */
+  public async startNewBlock(
+    numTxs: number,
+    globalVariables: GlobalVariables,
+    l1ToL2Messages: Fr[],
+    emptyTx: ProcessedTx,
+  ): Promise<ProvingTicket> {
+    logger.info(`Starting new block with ${numTxs} transactions`);
+    // we start the block by enqueueing all of the base parity circuits
+    let baseParityInputs: BaseParityInputs[] = [];
+    let l1ToL2MessagesPadded: Tuple<Fr, typeof NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP>;
+    try {
+      l1ToL2MessagesPadded = padArrayEnd(l1ToL2Messages, Fr.ZERO, NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP);
+    } catch (err) {
+      throw new Error('Too many L1 to L2 messages');
+    }
+    baseParityInputs = Array.from({ length: NUM_BASE_PARITY_PER_ROOT_PARITY }, (_, i) =>
+      BaseParityInputs.fromSlice(l1ToL2MessagesPadded, i),
+    );
+
+    //TODO:(@PhilWindle) Temporary until we figure out when to perform L1 to L2 insertions to make state consistency easier.
+    await Promise.resolve();
+
+    const promise = new Promise<ProvingResult>((resolve, reject) => {
+      this.provingState = new ProvingState(
+        numTxs,
+        resolve,
+        reject,
+        globalVariables,
+        l1ToL2MessagesPadded,
+        baseParityInputs.length,
+        emptyTx,
+      );
+    });
+
+    for (let i = 0; i < baseParityInputs.length; i++) {
+      this.enqueueJob(this.provingState!.stateIdentifier, PROVING_JOB_TYPE.BASE_PARITY, () =>
+        this.runBaseParityCircuit(baseParityInputs[i], i, this.provingState!.stateIdentifier),
+      );
+    }
+
+    const ticket: ProvingTicket = {
+      provingPromise: promise,
+    };
+    return ticket;
+  }
+
+  /**
+   * The interface to add a simulated transaction to the scheduler
+   * @param tx - The transaction to be proven
+   */
+  public async addNewTx(tx: ProcessedTx): Promise<void> {
+    if (!this.provingState) {
+      throw new Error(`Invalid proving state, call startNewBlock before adding transactions`);
+    }
```

Comment:
Possible alternative is to have the `ProvingOrchestrator` hang on to ticket, and have `startNewBlock` reject if the ticket's promise has not fulfilled.

Reply:
ProvingOrchestrator is dependent on being called sequentially. You wouldn't be able to build more than 1 block at a time as it would corrupt the MerkleTrees.